src/ship/GunManager.cpp

```diff
@@ -0,0 +1,519 @@
+		m_weapons.erase(iter);
+	}
+}
+
+bool GunManager::IsWeaponMounted(StringName hardpoint) const
+{
+	return std::find_if(m_weapons.begin(), m_weapons.end(), [&](const WeaponState &ws) {
+		return ws.mount->id == hardpoint;
+	}) != m_weapons.end();
+}
+
+void GunManager::RemoveGroupIndex(WeaponIndexSet &group, uint32_t index)
+{
+	// Mask containing all of the indices below index
+	WeaponIndexSet keep = WeaponIndexSet().set() >> (NUM_WEAPON_INDICES - index);
+	// Mask containing all of the indices above index
+	WeaponIndexSet mask = WeaponIndexSet().set() << (index + 1);
+
+	// Decrement all contained indices in the set above the removed index down by one
+	group = (group & keep) | ((group & mask) >> 1);
+}
+
+uint32_t GunManager::GetWeaponIndexForHardpoint(StringName hardpoint) const
+{
+	auto iter = std::find_if(m_weapons.begin(), m_weapons.end(), [&](const WeaponState &ws) {
+		return ws.mount->id == hardpoint;
+	});
+
+	if (iter == m_weapons.end()) {
+		return UINT32_MAX;
+	}
+
+	return std::distance(m_weapons.begin(), iter);
+}
+
+void GunManager::SetupDefaultGroups()
+{
+	m_groups.clear();
+
+	GroupState group = {};
+
+	for (size_t i = 0; i < m_weapons.size(); i++) {
+		group.weapons[i] = true;
+	}
+
+	m_groups.push_back(group);
+}
+
+void GunManager::SetAllGroupsFiring(bool firing)
+{
+	for (size_t idx = 0; idx < m_groups.size(); idx++) {
+		SetGroupFiring(idx, firing);
+	}
+
+	m_isAnyFiring = firing;
+}
+
+void GunManager::AssignWeaponToGroup(uint32_t numWeapon, uint32_t group)
+{
+	if (numWeapon > m_weapons.size())
+		return;
+
+	if (group >= m_groups.size()) {
+		m_groups.resize(group + 1);
+	}
+
+	WeaponState &ws = m_weapons[numWeapon];
+
+	uint32_t oldGroup = ws.group;
+	m_groups[oldGroup].weapons[numWeapon] = false;
+
+	// Prune empty groups when the last weapon is removed
+	if (m_groups[oldGroup].weapons.none()) {
+		RemoveGroup(oldGroup);
+	}
+
+	ws.group = group;
+	m_groups[group].weapons[numWeapon] = true;
+}
+
+void GunManager::RemoveGroup(uint32_t group)
+{
+	if (group == 0 || group >= m_groups.size())
+		return;
+
+	GroupState &gzero = m_groups[0];
+
+	for (size_t idx = 0; idx < m_weapons.size(); idx++) {
+		WeaponState &weapon = m_weapons[idx];
+
+		// Reassign weapons from this group to group zero
+		if (weapon.group == group) {
+			weapon.group = 0;
+			gzero.weapons[idx] = true;
+		}
+
+		// Reduce index of all groups above this
+		if (weapon.group > group) {
+			weapon.group--;
+		}
+	}
+
+	// Finally, clear the group from the list
+	m_groups.erase(m_groups.begin() + group);
+}
+
+const Body *GunManager::GetGroupTarget(uint32_t group)
+{
+	if (group >= m_groups.size())
+		return nullptr;
+
+	return m_groups[group].target;
+}
+
+void GunManager::SetGroupTarget(uint32_t group, const Body *target)
+{
+	if (group >= m_groups.size())
+		return;
+
+	GroupState &gs = m_groups[group];
+	// No weapons at all, ignore
+	if (!gs.weapons.any())
+		return;
```

Comment:
Given groups appear to get pruned when the last weapon is removed from them, is this check required in the current implementation?

(I'll discuss separately about design decisions regarding empty groups etc)

Reply:
This is technically unneeded, though I've not fully decided whether I want automatic group removal to remain a thing or if it should be an explicit action on the user's part.